# Practica # 1 - Tensorflow - Gradient Descent y Tensorboard

## Datos de Proyecto - Carga

Se cargará el Dataset del Proyecto del Curso **"Ciencia de Datos en Python"**; para referencia se coloca el listado de variables que contiene el Dataset:
* **SalePrice:** The property's sale price in dollars. This is the target variable that you're trying to predict.
* **OverallQual:** Overall material and finish quality, rates the overall material and finish of the house
       10	Very Excellent
       9	Excellent
       8	Very Good
       7	Good
       6	Above Average
       5	Average
       4	Below Average
       3	Fair
       2	Poor
       1	Very Poor
* **1stFlrSF:** First Floor square feet
* **TotRmsAbvGrd:** Total rooms above grade (does not include bathrooms)
* **YearBuilt:** Original construction date
* **LotFrontage:** Linear feet of street connected to property

In [9]:
#Importamos Libreria de Numpy para cargar los datos
import numpy as np

House_Data = np.load('proyecto_data/proyecto_training_data.npy')

##Verificamos que haya cargado la información, y que tenga la forma requerida
print('Rango Tensor:',House_Data.ndim)
print('Forma Tensor:',House_Data.shape)
print('Vemos Primer Registro:',House_Data[0])

Rango Tensor: 2
Forma Tensor: (1460, 6)
Vemos Primer Registro: [2.085e+05 7.000e+00 8.560e+02 8.000e+00 2.003e+03 6.500e+01]


## Datos de Entrenamiento

Consideremos lo siguiente:
* La variable Dependiente seguirá siendo **"SalesPrice"**
* La variable Independiente (FEATURE) será **"OverallQual"** que demostró una mayor correlación para el proyecto pasado.
* Tomaremos todo el Dataset para Entrenamiento, esta vez no separaremos el Dataset.

In [10]:
#Por Facilidad se manejaran tensores quitando los datos que no utilizaremos
##OverallQual - Variable Independiente (FEATURE)
x_train = House_Data[:,1]

##Valor de Ventas Real -  Variable Dependiente
##Al Igual que el Proyecto por facilidad dividiremos por 1000 el precio de las casas
y_real = House_Data[:,0]/1000

## TensorFlow - Definición Modelo

Importamos las librerias y ejecutamos los comandos requeridos para utilizar tensorflow.

In [11]:
import tensorflow as tf

print(tf.__version__)

2.1.0


In [12]:
if tf.__version__.startswith("2."):
  import tensorflow.compat.v1 as tf
  tf.compat.v1.disable_v2_behavior()
  tf.compat.v1.disable_eager_execution()
  print("Enabled compatitility to tf1.x")

Instructions for updating:
non-resource variables are not supported in the long term
Enabled compatitility to tf1.x


Creamos la Clase que almacenará la definición del Modelo

In [13]:
'''Clase que Definirá el Modelo a Entrenar'''
class Modelo:
    
    '''Metodo que define lo que se ejecuta al Instanciar la Clase'''
    def __init__(self):
        
        #Limpiamos cualquier variable o operación en el grafo 
        tf.reset_default_graph()
        
        #Variables que "Entrenará" el Modelo
        self.m = tf.get_variable("pendiente",dtype=tf.float32,shape=[],initializer=tf.zeros_initializer())
        self.b = tf.get_variable("intercepto",dtype=tf.float32,shape=[],initializer=tf.zeros_initializer())

        ##Definimos los Tensores para la Inicialización de Variables
        with tf.name_scope("Inicializacion"):
            init_global = tf.global_variables_initializer()
            init_local = tf.local_variables_initializer()
            self.init = tf.group(init_global,init_local)
    
    '''Metodo que define el Modelo de Predicción'''    
    def __call__(self,x):
        ##Definimos el Modelo de Predicción
        with tf.name_scope("Modelo"):
            #lo almacenamos en la propiedad de la clase self.y_hat para que pueda reutilizarse en los otros metodos
            #y no llamarlos varias veces - creando multiples instancias del modelo en tensorboard
            self.y_hat = tf.add(tf.multiply(self.m,x),self.b,name="Prediccion_Modelo")
        return self.y_hat
    
    '''Metodo que define la Función de Error del Modelo''' 
    def error_function(self,y,y_hat):
        ##Definimos la función de Error
        with tf.name_scope("Error"):
            #lo almacenamos en la propiedad de la clase self.error para que pueda reutilizarse en los otros metodos
            #y no llamarlos varias veces - creando multiples instancias del modelo en tensorboard
            self.error = 1/2*tf.reduce_mean(tf.math.square(y - y_hat) )
        return self.error
    
    '''Metodo que define como se actualizan las variables que se entrenan en el Modelo''' 
    def actualizar(self,x,y,learning_rate):
        #Quitamos esto - Para no crear multiples tensores con el modelo de predicción
        #y_hat = self(x) 
    
        #Quitamos esto - Para no crear multiples tensores con el modelo de Error
        #error = self.error_function(y,self.y_hat)
        
        #Calculo de la Gradiente - Para variables de Pendiente e Intercepto
        with tf.name_scope("Calculo_Gradiente"):
            gradiente = tf.gradients(self.error,[self.m, self.b])
        
        #Actualización de Variables del Modelo - Definición Tensores
        with tf.name_scope("Act_Parametros_Modelo"):
            actualizacion_m = tf.assign(self.m, self.m - learning_rate*gradiente[0] )
            actualizacion_b = tf.assign(self.b, self.b - learning_rate*gradiente[1] )
            actualizacion_parametros = tf.group(actualizacion_m,actualizacion_b)

        return actualizacion_parametros
    
    '''Metodo que define el Scalar para Tensorboard'''
    def summary(self,x,y):
        #Quitamos esto - Para no crear multiples tensores con el modelo de predicción
        #self.y_hat = self(x)
        
        #Quitamos esto - Para no crear multiples tensores con el modelo de Error
        #error = self.error_function(y,self.y_hat)
        
        #Creamos el Tensor con la Información del Scalar
        MSE_summary = tf.summary.scalar(name='MSE_Function', tensor=self.error)
        return MSE_summary
        

Definimos la Función de Entrenamiento

In [14]:
import time

def Entrenamiento(learning_rate,epochs,imprimir_cada,x,y):
    #Instanciamos la Clase que define el Modelo
    modelo = Modelo()

    #string de configuración para almacenamiento del Grafo
    string_conf= time.strftime("%Y%m%d") + "_"+time.strftime("%H%M%S")+"_Batch_GradientDescent_Lr="+str(learning_rate)

    #Definimos los Placeholders que almacenarán los datos de Entrenamiento del Modelo
    tensor_x = tf.placeholder(tf.float32,[x.shape[0]],name="tensor_x")
    tensor_y = tf.placeholder(tf.float32,[x.shape[0]],name="tensor_y")

    #Llamamos a la función que define el modelo de Predicción
    prediccion = modelo(tensor_x)

    #Llamamos a la función que define la Función de Error (Costo)
    funcion_error = modelo.error_function(prediccion,tensor_y)

    #Llamamos a la función que define los tensores para actualizar las variables del Modelo
    actualizacion_parametros = modelo.actualizar(tensor_x,tensor_y,learning_rate)

    #llamamos a la función que define el Scalar a Reportar en Tensorboard
    MSE_summary = modelo.summary(tensor_x,tensor_y)

    #Iniciamos con una Sesión Monitoreada - Para el Entrenamiento del Modelo
    with tf.train.MonitoredSession() as session:

        # Inicialización de variables Globales y locales
        session.run(modelo.init)

        #Escritura del Grafo para Tensorboard
        writer = tf.summary.FileWriter('./graphs/'+string_conf, session.graph)

        #Creamos el Diccionario que le dará valor a los Placeholders
        feed_dict = {tensor_x:x, tensor_y:y}

        #Entrenamiento del Modelo
        for step in range(epochs):

            #Ejecutamos el Grafo - Actualización de Datos
            session.run(actualizacion_parametros,feed_dict=feed_dict)

            #Ejecutamos el Grafo - Predicción del Modelo
            predicciones = session.run(prediccion,feed_dict=feed_dict)

            #Ejecutamos el Grafo - Resultado de Variables Entrenadas
            pendiente_final,intercepto_final = session.run([modelo.m,modelo.b],feed_dict=feed_dict)

            #Ejecutamos el Grafo - Datos Escalar
            summary = session.run(MSE_summary,feed_dict=feed_dict)

            #Almacenamos datos de Scalar - Para Tensorboard
            writer.add_summary(summary, step)

            #Imprimimos resultados de Variables por entrenamiento - Cada N Entrenamientos
            if (step+1) % imprimir_cada == 0:
                print("Prueba {} - Parámetros finales: m={}  b={}".format(step+1,pendiente_final,intercepto_final))

        # Si queremos imprimir la función de Error - Despues del Entrenamiento 
        final_error = session.run(funcion_error,feed_dict=feed_dict)
        print("Error - Despues de Entrenamiento={}".format(final_error))

Probamos el Modelo solo para asegurar que no haya errores. Por el momento es con valores pequeños de prueba

In [16]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.001

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=100

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 50

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 50 - Parámetros finales: m=25.88538360595703  b=3.795400619506836
Prueba 100 - Parámetros finales: m=29.268674850463867  b=4.078108310699463
Error - Despues de Entrenamiento=1431.6644287109375


## Tensorboard - Grafo

Vemos como quedo nuestro Grafo en Tensorboard, de acuerdo al modelo definido.

<img src="assets/Grafo_Tensorboard.png">

## Tensorboard - Scalar

El Error obtenido por la función MSE (Mean Squared Error) se puede visualizar en tensorbord de la siguiente manera:

<img src="assets/Scalar_Tensorboard.png">

Vemos tambien como en Tensorboard se muestra la prueba con el string de configuración definido, y la diferencia en algunas pruebas ejecutadas en las correciones del modelo.

<img src="assets/Tensorboard_StringConf.png">

## Pruebas de Entrenamiento para Ajuste Parametros Modelo

Haremos pruebas para ir ajustando los hyperparametros para el correcto funcionamiento de nuestro modelo.

Anteriormente **(Modelo 1)** ejecutamos el entrenamiento con un learning rate de **0.001**, pero solamente ejecutamos **100** pruebas consiguiendo un error de **1431**. Variaremos los parametros para ir consiguiendo mejores resultados.

### Modelo 2

No cambiaremos el Learning rate, pero ejecutaremos al menos unas 1000 pruebas.

In [19]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.001

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=1000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=30.062368392944336  b=2.2587428092956543
Prueba 1000 - Parámetros finales: m=30.42473793029785  b=-0.06219841539859772
Error - Despues de Entrenamiento=1406.33251953125


Conseguimos un mejor resultado **(1406)**; pero veo que no es suficiente. Veamos como cambió la grafica en Tensorboard, comparando con el modelo anterior.

<img src="assets/Scalar_Tensorboard_Modelo2.png">


### Modelo 3

No cambiaremos el Learning rate, pero ejecutaremos más pruebas, al menos unas 3000.

In [23]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.001

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=3000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=30.062368392944336  b=2.2587428092956543
Prueba 1000 - Parámetros finales: m=30.42473793029785  b=-0.06219841539859772
Prueba 1500 - Parámetros finales: m=30.778564453125  b=-2.3284294605255127
Prueba 2000 - Parámetros finales: m=31.12404441833496  b=-4.541244029998779
Prueba 2500 - Parámetros finales: m=31.46140480041504  b=-6.701898574829102
Prueba 3000 - Parámetros finales: m=31.790790557861328  b=-8.811623573303223
Error - Despues de Entrenamiento=1367.0936279296875


Conseguimos un mejor resultado **(1367)**; pero veo que por el ser el Learning rate muy pequeño vamos avanzando muy poco. Veamos como cambió la grafica en Tensorboard, comparando con el modelo anterior.

<img src="assets/Scalar_Tensorboard_Modelo3.png">


### Modelo 4

Al ver poco avance con el Learning rate, cambiaremos a un learning rate de 0.01 siempre dejando pruebas 3000.

In [24]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.01

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=3000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=33.033206939697266  b=-16.769079208374023
Prueba 1000 - Parámetros finales: m=35.66579818725586  b=-33.630496978759766
Prueba 1500 - Parámetros finales: m=37.73958969116211  b=-46.91284942626953
Prueba 2000 - Parámetros finales: m=39.37319564819336  b=-57.375877380371094
Prueba 2500 - Parámetros finales: m=40.6600456237793  b=-65.61799621582031
Prueba 3000 - Parámetros finales: m=41.67375564575195  b=-72.11067199707031
Error - Despues de Entrenamiento=1194.6534423828125


Conseguimos un mejor resultado **(1194)**; se consiguió un mejor resultado incrementando el learning rate. Veamos como cambió la grafica en Tensorboard, comparando con el modelo anterior.

<img src="assets/Scalar_Tensorboard_Modelo4.png">


##### Modelo 5

Volveremos a cambiar el Learning rate, dejandolo ahora en 0.1, y por el momento siempre con 3000 pruebas.

In [25]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.1

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=3000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=nan  b=nan
Prueba 1000 - Parámetros finales: m=nan  b=nan
Prueba 1500 - Parámetros finales: m=nan  b=nan
Prueba 2000 - Parámetros finales: m=nan  b=nan
Prueba 2500 - Parámetros finales: m=nan  b=nan
Prueba 3000 - Parámetros finales: m=nan  b=nan
Error - Despues de Entrenamiento=nan


Aquí vemos que con este Learning rate, no se encontró divergencia. Debemos probar un learning rate menor.


##### Modelo 6

El Learning rate anterior fue muy alto, lo rebajaremos a 0.05 y por el momento siempre con 3000 pruebas.

In [26]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.05

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=3000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=-110.8829345703125  b=-89.3142318725586
Prueba 1000 - Parámetros finales: m=-729.5380249023438  b=-207.72073364257812
Prueba 1500 - Parámetros finales: m=-3903.27099609375  b=-709.8502197265625
Prueba 2000 - Parámetros finales: m=-20107.7109375  b=-3241.874267578125
Prueba 2500 - Parámetros finales: m=-102818.453125  b=-16156.212890625
Prueba 3000 - Parámetros finales: m=-524986.5  b=-82069.34375
Error - Despues de Entrenamiento=5656783880192.0


Aquí vemos que con este Learning rate tampoco funcionó correctamente, mucha divergencia, consiguiendo valores de error muy altos. Veamos como cambió la grafica en Tensorboard, comparando con los modelos anteriores. 

<img src="assets/Scalar_Tensorboard_Modelo6.png">

Debemos probar un learning rate menor.


##### Modelo 7

El Learning rate anterior fue muy alto, lo rebajaremos a 0.04 y por el momento siempre con 3000 pruebas.

In [27]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.04

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=3000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=39.3773307800293  b=-57.402374267578125
Prueba 1000 - Parámetros finales: m=43.104515075683594  b=-81.27448272705078
Prueba 1500 - Parámetros finales: m=44.53873062133789  b=-90.46044921875
Prueba 2000 - Parámetros finales: m=45.09061050415039  b=-93.99517822265625
Prueba 2500 - Parámetros finales: m=45.302982330322266  b=-95.3553466796875
Prueba 3000 - Parámetros finales: m=45.38468933105469  b=-95.87870025634766
Error - Despues de Entrenamiento=1180.4698486328125


Conseguimos un mejor resultado **(1180.46)**; se consiguió un mejor resultado con el learning rate de 0.04 que con los modelos anteriores. Veamos como cambió la grafica en Tensorboard, comparando con los modelos de 0.001, 0.01.

<img src="assets/Scalar_Tensorboard_Modelo7.png">


##### Modelo 8

Mantendremos el learning rate de 0.04, pero ahora aumentaremos las pruebas a 4000 para ver si mejora el modelo con más pruebas.

In [28]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.04

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=4000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 1000

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 1000 - Parámetros finales: m=43.104515075683594  b=-81.27448272705078
Prueba 2000 - Parámetros finales: m=45.09061050415039  b=-93.99517822265625
Prueba 3000 - Parámetros finales: m=45.38468933105469  b=-95.87870025634766
Prueba 4000 - Parámetros finales: m=45.42823028564453  b=-96.15756225585938
Error - Despues de Entrenamiento=1180.4676513671875


Se podría decir que se consiguió los mismos resultados que con 3000 pruebas **(1180.46)**; hubo una mejora en centecimos de decimal, pero a mi parecer no valdría la pena. Veamos como cambió la grafica en Tensorboard, comparando con los modelos de 0.001, 0.01, 0.04 (3000 Pruebas).

<img src="assets/Scalar_Tensorboard_Modelo8.png">

##### Modelo 9

Mantendremos el learning rate de 0.04, pero ahora reduciremos a 2000 para ver si este número es suficiente para conseguir los resultados conseguido en el modelo 7.

In [29]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.04

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=2000

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=39.3773307800293  b=-57.402374267578125
Prueba 1000 - Parámetros finales: m=43.104515075683594  b=-81.27448272705078
Prueba 1500 - Parámetros finales: m=44.53873062133789  b=-90.46044921875
Prueba 2000 - Parámetros finales: m=45.09061050415039  b=-93.99517822265625
Error - Despues de Entrenamiento=1180.5867919921875


Con menos pruebas, no se consiguio un mejor resultado que el modelo 7. Tiene un valor cercano pero se puede mejorar con mas pruebas. No se muestra la gráfica de tensorboard, porque a nivel visual no hay mayor diferencia a la grafica anterior.

##### Modelo 10

Mantendremos el learning rate de 0.04, pero ahora aumentaremos las pruebas a 2500 para ver si este número es suficiente para conseguir los resultados conseguido en el modelo 7.

In [30]:
#Usaremos el mismo Learning rate del Proyecto
learning_rate = 0.04

#Definimos cuantas iteraciones de Entrenamiento haremos
epochs=2500

#Definimos cada cuanto vamos a imprimir
imprimir_cada = 500

#Ejecutamos la función de Entrenamiento para los parametros definidos
Entrenamiento(learning_rate,epochs,imprimir_cada,x_train,y_real)

INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
Prueba 500 - Parámetros finales: m=39.3773307800293  b=-57.402374267578125
Prueba 1000 - Parámetros finales: m=43.104515075683594  b=-81.27448272705078
Prueba 1500 - Parámetros finales: m=44.53873062133789  b=-90.46044921875
Prueba 2000 - Parámetros finales: m=45.09061050415039  b=-93.99517822265625
Prueba 2500 - Parámetros finales: m=45.302982330322266  b=-95.3553466796875
Error - Despues de Entrenamiento=1180.484619140625


Con 2500 pruebas, casí se consiguio el mismo resultado que el modelo 7. Tiene un valor cercano pero se puede mejorar con mas pruebas. No se muestra la gráfica de tensorboard, porque a nivel visual no hay mayor diferencia a la grafica del modelo 8.

## Conclusión

Despues de efectuar muchas pruebas, variando número de pruebas, learning rate; el Modelo que más se ajusta sería **Modelo #7**. Los datos serían:
* Learning rate = 0.04
* EPOCHS = 3000
* Pendiente (m) = 45.38468933105469
* Intercepto (b) =-95.87870025634766
* MSE al Finalizar la Prueba = 1180.46

La función para la variable Overall quedaria de la siguiente manera: y = 45.38468933105469x -95.87870025634766
